### PR TITLE
[MLIR] Use prebuilt MLIR when specified with -DNGRAPH_USE_PREBUILT_MLIR

### PR DESCRIPTION
**Pull request by pruthviIntel**
```
pthoreho@aipg-ra-clx-ap-04:~/ngraph/build_mlir_prebuilt$ cmake ../ -DCMAKE_INSTALL_PREFIX=~/ngraph_dist -DNGRAPH_ONNX_IMPORT_ENABLE=ON -DNGRAPH_ONNXIFI_ENABLE=ON -DNGRAPH_CPU_ENABLE=TRUE -DNGRAPH_USE_PREBUILT_LLVM=TRUE -DNGRAPH_MLIR_ENABLE=TRUE -DNGRAPH_DEX_ONLY=ON -DNGRAPH_USE_PREBUILT_MLIR=True -DMLIR_LLVM_PREBUILT_PATH="/localdisk/pthoreho/"

-- Found Git: /usr/bin/git (found version "2.17.1")
-- Current hash 53a74fb122283494641cdcb5f8fde4cfc08c0fae
-- NGRAPH_VERSION 0.27.1-rc.0+53a74fb
-- NGRAPH_VERSION_SHORT 0.27.1-rc.0
-- NGRAPH_WHEEL_VERSION 0.27.1rc0
-- NGRAPH_API_VERSION 0.27
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- NGRAPH_FORWARD_CMAKE_ARGS -DCMAKE_C_COMPILER=/usr/bin/cc;-DCMAKE_CXX_COMPILER=/usr/bin/c++;-DCMAKE_CXX_STANDARD:STRING=14;-DCMAKE_CXX_STANDARD_REQUIRED:BOOL=ON;-DCMAKE_CXX_EXTENSIONS:BOOL=OFF;-DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=ON;-DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON;-DCMAKE_BUILD_TYPE=Release
-- NGRAPH_CXX_STANDARD:                  14
-- NGRAPH_UNIT_TEST_ENABLE:              ON
-- NGRAPH_DOC_BUILD_ENABLE:              OFF
-- NGRAPH_TOOLS_ENABLE:                  ON
-- NGRAPH_CPU_ENABLE:                    ON
-- NGRAPH_USE_LEGACY_MKLDNN:             ON
-- NGRAPH_MLIR_ENABLE:                   ON
-- NGRAPH_INTERPRETER_ENABLE:            ON
-- NGRAPH_NOP_ENABLE:                    ON
-- NGRAPH_GENERIC_CPU_ENABLE:            OFF
-- NGRAPH_DEBUG_ENABLE:                  OFF
-- NGRAPH_DEPRECATED_ENABLE:             OFF
-- NGRAPH_ONNX_IMPORT_ENABLE:            ON
-- NGRAPH_DEX_ONLY:                      ON
-- NGRAPH_ENABLE_CPU_CONV_AUTO:          ON
-- NGRAPH_CODE_COVERAGE_ENABLE:          OFF
-- NGRAPH_LIB_VERSIONING_ENABLE:         OFF
-- NGRAPH_PYTHON_BUILD_ENABLE:           OFF
-- NGRAPH_USE_PREBUILT_LLVM:             ON
-- NGRAPH_USE_PREBUILT_MLIR:             ON
-- NGRAPH_PLAIDML_ENABLE:                OFF
-- NGRAPH_DISTRIBUTED_ENABLE:            OFF
-- NGRAPH_JSON_ENABLE:                   ON
-- NGRAPH_STATIC_LIB_ENABLE:             OFF
-- NGRAPH_INTERPRETER_STATIC_LIB_ENABLE: OFF
-- NGRAPH_CPU_STATIC_LIB_ENABLE:         OFF
-- NGRAPH_PLAIDML_STATIC_LIB_ENABLE:     OFF
-- NGRAPH_DYNAMIC_COMPONENTS_ENABLE:     ON
-- NGRAPH_NATIVE_ARCH_ENABLE:            ON
-- NGRAPH_TBB_ENABLE:                    ON
-- Installation directory: /nfs/pdx/home/pthoreho/ngraph_dist
-- Compile Flags:  -DEIGEN_MPL2_ONLY -DTBB_USE_THREADING_TOOLS -O2 -Wformat -Wformat-security -D_FORTIFY_SOURCE=2 -fstack-protector-strong -march=native -DNGRAPH_CPU_ENABLE -DNGRAPH_USE_LEGACY_MKLDNN -DNGRAPH_MLIR_ENABLE -DNGRAPH_DYNAMIC_COMPONENTS_ENABLE
-- Shared Link Flags:  -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now
-- CMAKE_CXX_FLAGS_RELEASE -O3 -DNDEBUG
-- CMAKE_CXX_FLAGS_DEBUG -O0 -g
-- Linker detection: GNU ld
-- Found LLVM 10.0.0git
-- Using modules in: /localdisk/pthoreho/mlir_project/llvm-projects/build/lib/cmake/llvm
-- LLVM RTTI is ON
-- MLIR headers at: /localdisk/pthoreho//mlir_project/llvm-projects/llvm/projects/mlir/include;/localdisk/pthoreho//mlir_project/llvm-projects/build/projects/mlir/include
-- LLVM headers at: /nfs/pdx/home/pthoreho/ngraph/build_mlir/mlir_project/llvm-projects/llvm/include;/nfs/pdx/home/pthoreho/ngraph/build_mlir/mlir_project/llvm-projects/build/include
-- Configuring done
-- Generating done
-- Build files have been written to: /ec/pdx/disks/aipg_lab_home_pool_02/pthoreho/ngraph/build_mlir_prebuilt/tbb
Scanning dependencies of target ext_tbb
[ 12%] Creating directories for 'ext_tbb'
[ 25%] Performing download step (git clone) for 'ext_tbb'
Cloning into 'tbb-src'...
Checking out files: 100% (1005/1005), done.
Note: checking out '2019_U3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 3147923 Committing TBB 2019 Update 3 source code
[ 37%] No patch step for 'ext_tbb'
[ 50%] No update step for 'ext_tbb'
[ 62%] No configure step for 'ext_tbb'
[ 75%] No build step for 'ext_tbb'
[ 87%] No install step for 'ext_tbb'
[100%] Completed 'ext_tbb'
[100%] Built target ext_tbb
-- Building Intel TBB: /usr/bin/make -j192 compiler=gcc tbb_build_dir=/nfs/pdx/home/pthoreho/ngraph/build_mlir_prebuilt/tbb_build tbb_build_prefix=tbb stdver=c++14
-- Found TBB and imported target TBB::tbb
-- TBB so version: 2
-- MLIR Ops link flag: -Llib -Wl,--whole-archive,/nfs/pdx/home/pthoreho/ngraph/build_mlir/mlir_project/llvm-projects/build/./lib/libMLIRAffineOps.a,/nfs/pdx/home/pthoreho/ngraph/build_mlir/mlir_project/llvm-projects/build/./lib/libMLIRStandardOps.a,--no-whole-archive
-- NGRAPH_FLUID_ENABLE:                  ON
-- NGRAPH_FLUID_TEST_ENABLE:             OFF
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Building CPU backend with MLIR
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- PlaidML not enabled; not compiling ngraph-to-plaidml
-- tools enabled
-- unit tests enabled
-- MLIR LIT tests enabled
-- Adding unit test for backend INTERPRETER
-- Adding unit test for backend CPU
-- Configuring done
-- Generating done
-- Build files have been written to: /nfs/pdx/home/pthoreho/ngraph/build_mlir_prebuilt
pthoreho@aipg-ra-clx-ap-04:~/ngraph/build_mlir_prebuilt$ 

```

`cmake ../ -DCMAKE_INSTALL_PREFIX=~/ngraph_dist -DNGRAPH_ONNX_IMPORT_ENABLE=ON -DNGRAPH_ONNXIFI_ENABLE=ON -DNGRAPH_CPU_ENABLE=TRUE -DNGRAPH_USE_PREBUILT_LLVM=TRUE -DNGRAPH_MLIR_ENABLE=TRUE -DNGRAPH_DEX_ONLY=ON -DNGRAPH_USE_PREBUILT_MLIR=True -DMLIR_LLVM_PREBUILT_PATH="/localdisk/pthoreho/"`

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -429,6 +432,7 @@ endif()
 
 if (NGRAPH_MLIR_ENABLE)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DNGRAPH_MLIR_ENABLE")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DNGRAPH_USE_PREBUILT_MLIR")
```
Comment (nmostafa) [inline CMakeLists.txt]: no need to set CXX flags, we won't check that in the source code.

File: cmake/external_mlir_prebuilt.cmake [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ******************************************************************************
+
+include(ExternalProject)
+
+ExternalProject_Add(
+    ext_mlir
+    PREFIX mlir
+    URL ${MLIR_LLVM_TARBALL_URL}
```
Comment (nmostafa) [inline cmake/external_mlir_prebuilt.cmake]: Where will these flags come from ? CMAKE options ? If so, please add them as CMAKE options +status message in top file.

File: cmake/external_mlir_prebuilt.cmake [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    UPDATE_COMMAND ""
+    DOWNLOAD_NO_PROGRESS TRUE
+    EXCLUDE_FROM_ALL TRUE
+    )
+
+ExternalProject_Get_Property(ext_mlir SOURCE_DIR)
+set(INSTALL_DIR ${SOURCE_DIR})
+
+# MLIR environment variables. Some of them are used by LIT tool.
+set(MLIR_LLVM_ROOT ${INSTALL_DIR}/llvm-projects)
```
Comment (nmostafa) [inline cmake/external_mlir_prebuilt.cmake]: There's duplication with non-prebuilt file here. Can we unify them ?

File: CMakeLists.txt [hunk 4]
```diff
@@ -579,8 +583,10 @@ if(NGRAPH_CPU_ENABLE)
         include(cmake/external_mkldnn_v1.cmake)
     endif()
 endif()
-if (NGRAPH_MLIR_ENABLE)
+if (NGRAPH_MLIR_ENABLE AND NOT NGRAPH_USE_PREBUILT_MLIR)
```
Comment (nmostafa) [inline CMakeLists.txt]: nit: 
```suggestion
if (NGRAPH_MLIR_ENABLE)
if (NGRAPH_USE_PREBUILT_MLIR)
include(cmake/external_mlir_prebuilt.cmake).
else()
include(cmake/external_mlir.cmake)
endif()
endif()
```

File: cmake/external_mlir_prebuilt.cmake (hunk 3 above)
Comment (nmostafa) [reply cmake/external_mlir_prebuilt.cmake]: Maybe move the prebuilt check down here instead.

**Comment by pruthviIntel** [comment]
@nmostafa  I made some changes can u take a look again? :)

File: cmake/external_mlir.cmake [hunk 5]
```diff
@@ -24,7 +24,13 @@ set(MLIR_LLVM_COMMIT_ID 75990950)
 set(MLIR_COMMIT_ID 5e64e536)
 
 # MLIR environment variables. Some of them are used by LIT tool.
-set(MLIR_PROJECT_ROOT ${CMAKE_CURRENT_BINARY_DIR}/mlir_project)
+
+if (NGRAPH_USE_PREBUILT_MLIR)
+    set(MLIR_PROJECT_ROOT ${MLIR_LLVM_PREBUILT_PATH}/mlir_project)
```
Comment (dcaballe) [inline cmake/external_mlir.cmake]: is `mlir_project` necessary? That's a nGraph directory. We should be able to use this flag to compile nGraph with a vanilla llvm/mlir repo. Similar to what PlaidML is doing.

**Pull request merged**